rust_src/src/eval.rs

```diff
@@ -948,4 +951,118 @@ fn run_hook_with_args_internal(
     }
 }
 
+/// Call first argument as a function, passing remaining arguments to it.
+/// Return the value that function returns.
+/// Thus, (funcall \\='cons \\='x \\='y) returns (x . y).
+/// usage: (funcall FUNCTION &rest ARGUMENTS)
+#[allow(unused_assignments)]
+#[lisp_fn]
+pub fn funcall(args: &mut [LispObject]) -> LispObject {
+    unsafe { maybe_quit() };
+
+    // Increment the lisp eval depth
+    let mut current_thread = ThreadState::current_thread();
+    current_thread.m_lisp_eval_depth = current_thread.m_lisp_eval_depth + 1;
+
+    unsafe {
+        if current_thread.m_lisp_eval_depth > globals.max_lisp_eval_depth {
+            if globals.max_lisp_eval_depth < 100 {
+                globals.max_lisp_eval_depth = 100;
+            }
+
+            if current_thread.m_lisp_eval_depth > globals.max_lisp_eval_depth {
+                error!("Lisp nesting exceeds `max-lisp-eval-depth'");
+            }
+        }
+    }
+
+    // The first element in args is the called function.
+    let numargs = args.len() - 1;
+
+    let count = if numargs > 0 {
+        unsafe { record_in_backtrace(args[0], &mut args[1], numargs as isize) }
+    } else {
+        unsafe { record_in_backtrace(args[0], ptr::null_mut(), numargs as isize) }
+    };
+
+    unsafe { maybe_gc() };
+
+    unsafe {
+        if globals.debug_on_next_call {
+            do_debug_on_call(Qlambda, count);
+        }
+    }
+
+    unsafe { check_cons_list() };
+
+    let mut val = Qnil;
+
+    let original_fun = args[0];
+
+    loop {
+        if original_fun.is_nil() {
+            xsignal!(Qvoid_function, original_fun);
+        }
+
+        // Optimize for no indirection.
+        let fun = if original_fun.is_symbol() {
+            original_fun.as_symbol_or_error().get_indirect_function()
+        } else {
+            original_fun
+        };
+
+        if fun.is_nil() {
+            xsignal!(Qvoid_function, original_fun);
+        }
+
+        if fun.is_subr() {
+            val = unsafe {
+                funcall_subr(
+                    fun.as_subr_or_error().as_mut(),
+                    numargs as isize,
+                    args[1..].as_mut_ptr(),
+                )
+            };
+        } else if unsafe { COMPILEDP(fun) || MODULE_FUNCTIONP(fun) } {
+            val = unsafe { funcall_lambda(fun, numargs as isize, args[1..].as_mut_ptr()) };
+        } else if let Some(cons) = fun.as_cons() {
+            let funcar = cons.car();
+
+            if !funcar.is_symbol() {
+                xsignal!(Qinvalid_function, original_fun);
+            }
+
+            if funcar.eq(Qlambda) || funcar.eq(Qclosure) {
+                val = unsafe { funcall_lambda(fun, numargs as isize, args[1..].as_mut_ptr()) };
+            } else if funcar.eq(Qautoload) {
+                Fautoload_do_load(fun, original_fun, Qnil);
+                unsafe {
+                    check_cons_list();
+                }
+                continue;
```

Comment:
We have the same else twice. How do we debug this if it fails? The C code is cleaner. Can we simplify the Rust?

Reply:
Is this cleaner?

```rust
    if let Some(mut f) = fun.as_subr() {
        val = unsafe { funcall_subr(f.as_mut(), numargs as isize, args[1..].as_mut_ptr()) };
    } else if unsafe { COMPILEDP(fun) || MODULE_FUNCTIONP(fun) } {
        val = unsafe { funcall_lambda(fun, numargs as isize, args[1..].as_mut_ptr()) };
    } else if let Some(cons) = fun.as_cons() {
        let funcar = cons.car();

        if !funcar.is_symbol() {
            xsignal!(Qinvalid_function, original_fun);
        }

        if !funcar.eq(Qlambda) && !funcar.eq(Qclosure) && !funcar.eq(Qautoload) {
            xsignal!(Qinvalid_function, original_fun);
        }

        if funcar.eq(Qautoload) {
            Fautoload_do_load(fun, original_fun, Qnil);
            unsafe { check_cons_list() };
            continue;
        }

        val = unsafe { funcall_lambda(fun, numargs as isize, args[1..].as_mut_ptr()) };
    } else {
        xsignal!(Qinvalid_function, original_fun);
    }

```

Reply:
Reimplemented in a new way.